File: receipt_dynamo/receipt_dynamo/amounts.py
"""Receipt amount parsing helpers shared across receipt services."""

from __future__ import annotations

import re
from typing import Any

_CURRENCY_SYMBOLS = r"$€£¥₹"

# ---------------------------------------------------------------------------
# Shared summary-line keyword patterns.
#
# These are the canonical copies of the keyword patterns used to recognize
# receipt summary rows ("Total:", "BALANCE DUE", "Subtotal", ...). They are
# defined here (rather than in receipt_upload) because receipt_dynamo sits at
# the bottom of the dependency graph, so both receipt_upload's amount
# classifier and receipt_dynamo's ReceiptSummary printed-total fallback can
# import them without forking the regexes.
# ---------------------------------------------------------------------------

TOTAL_KEYWORD_RE = re.compile(
    r"\b(total|amount\s+due|balance|authorized)\b", re.I
)
SUBTOTAL_KEYWORD_RE = re.compile(r"\bsub[-\s]?total\b", re.I)
TAX_KEYWORD_RE = re.compile(r"\b(tax|vat)\b", re.I)
NON_PAYMENT_SUMMARY_RE = re.compile(
    r"\b("
    r"savings?|discounts?|refunds?|returns?|coupons?|promos?|"
    r"promotion|rewards?|loyalty|cash\s+back|cashback|store\s+credit"
    r")\b",
    re.I,
)
# Tokens that turn a "total" row into a non-monetary count/summary row
# (e.g. "TOTAL NUMBER OF ITEMS SOLD"), which must NOT be read as a
# grand-total row.
GRAND_TOTAL_DISQUALIFIER_TOKENS = frozenset(
    {
        "number",
        "items",
        "item",
        "qty",
        "quantity",
        "count",
        "sold",
        "transactions",
        "transaction",
        "pieces",
        "units",
        "lines",
        "savings",
        "saved",
    }
)


def is_grand_total_line(line_text: str) -> bool:
    """Return whether joined line text reads as a grand-total row.

    A grand-total row carries a total/balance/amount-due keyword, is not a
    subtotal or tax row, is not a non-payment summary row (savings,
    discounts, ...), and is not an item-count row ("TOTAL NUMBER OF ITEMS
    SOLD").
    """
    if not line_text:
        return False
    if not TOTAL_KEYWORD_RE.search(line_text):
        return False
    if SUBTOTAL_KEYWORD_RE.search(line_text):
        return False
    if TAX_KEYWORD_RE.search(line_text):
        return False
    if NON_PAYMENT_SUMMARY_RE.search(line_text):
        return False
    tokens = {
        token for token in re.split(r"[^a-z]+", line_text.lower()) if token
    }
    return not (tokens & GRAND_TOTAL_DISQUALIFIER_TOKENS)


def parse_receipt_amount(text: Any) -> float | None:
    """Parse receipt currency/amount text without changing the OCR text.

    Handles US-style amounts (``1,234.56``), decimal-comma amounts
    (``8,82``), European grouped amounts (``1.234,56``), and negative
    accounting forms.
    """
    if text is None:
        return None

    raw = str(text).strip()
    if not raw:
        return None

    is_negative = raw.startswith("(") and raw.endswith(")")
    cleaned = re.sub(f"[{re.escape(_CURRENCY_SYMBOLS)}\\s]", "", raw)

    if cleaned.startswith("(") and cleaned.endswith(")"):
        cleaned = cleaned[1:-1]

    if cleaned.endswith("-"):
        is_negative = True
        cleaned = cleaned[:-1]

    if cleaned.startswith("-"):
        is_negative = True
        cleaned = cleaned[1:]

    cleaned = re.sub(r"[^0-9,.-]", "", cleaned)
    if not cleaned or not re.search(r"\d", cleaned):
        return None

    normalized = _normalize_decimal_separators(cleaned)
    if normalized is None:
        return None

    try:
        value = float(normalized)
    except ValueError:
        return None

    return -value if is_negative else value


def looks_like_receipt_amount(text: Any) -> bool:
    """Return whether text has receipt amount punctuation/symbol context."""
    if text is None:
        return False
    raw = str(text).strip()
    if not raw:
        return False
    return bool(
        re.search(f"[{re.escape(_CURRENCY_SYMBOLS)}]", raw)
        or re.fullmatch(r"-?\(?\d{1,3}(,\d{3})+(\.\d{2})?\)?-?", raw)
        or re.fullmatch(r"-?\(?\d+([.,]\d{2})\)?-?", raw)
    )


def _normalize_decimal_separators(cleaned: str) -> str | None:
    """Normalize supported comma/dot conventions to Python float syntax."""
    if "," not in cleaned:
        return cleaned

    last_comma = cleaned.rfind(",")
    last_dot = cleaned.rfind(".")

    if last_dot >= 0 and last_comma > last_dot:
        # European grouping: 1.234,56 -> 1234.56
        if re.fullmatch(r"\d{1,3}(\.\d{3})+,\d{2}", cleaned):
            return cleaned.replace(".", "").replace(",", ".")
        return cleaned.replace(".", "").replace(",", ".")

    if last_dot >= 0:
        # US grouping: 1,234.56 -> 1234.56
        return cleaned.replace(",", "")

    if re.fullmatch(r"\d{1,3}(,\d{3})+", cleaned):
        # Thousands-only grouping: 1,234 -> 1234
        return cleaned.replace(",", "")

    if re.fullmatch(r"\d+,\d{2}", cleaned):
        # Decimal comma: 8,82 -> 8.82
        return cleaned.replace(",", ".")

    # Ambiguous comma usage; preserve historical behavior as thousands.
    return cleaned.replace(",", "")

File: receipt_dynamo_rs/src/amounts.rs
//! Receipt amount parsing helpers shared with Python `receipt_dynamo.amounts`.

use std::sync::LazyLock;

use regex::Regex;

const CURRENCY_SYMBOLS: &str = r"$€£¥₹";

static TOTAL_KEYWORD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)\b(total|amount\s+due|balance|authorized)\b").unwrap());
static SUBTOTAL_KEYWORD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)\bsub[-\s]?total\b").unwrap());
static TAX_KEYWORD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)\b(tax|vat)\b").unwrap());
static NON_PAYMENT_SUMMARY_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"(?i)\b(savings?|discounts?|refunds?|returns?|coupons?|promos?|promotion|rewards?|loyalty|cash\s+back|cashback|store\s+credit)\b",
    )
    .unwrap()
});
static TENDER_KEYWORD_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"(?i)\b(tender(?:ed)?|cash|change)\b").unwrap());
static DECIMAL_AMOUNT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"-?\(?\d+([.,]\d{2})\)?-?").unwrap());
static GROUPED_AMOUNT_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"-?\(?\d{1,3}(,\d{3})+(\.\d{2})?\)?-?").unwrap());
static HAS_DIGIT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\d").unwrap());
static CURRENCY_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(&format!("[{}]", regex::escape(CURRENCY_SYMBOLS))).unwrap());
static STRIP_CURRENCY_WS: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(&format!(r"[{}\s]", regex::escape(CURRENCY_SYMBOLS))).unwrap());
static NON_NUMERIC: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^0-9,.-]").unwrap());
static TOKEN_SPLIT: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[^a-z]+").unwrap());
static EURO_GROUPED: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\d{1,3}(\.\d{3})+,\d{2}$").unwrap());
static THOUSANDS_ONLY: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"^\d{1,3}(,\d{3})+$").unwrap());
static DECIMAL_COMMA: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"^\d+,\d{2}$").unwrap());

const GRAND_TOTAL_DISQUALIFIER_TOKENS: &[&str] = &[
    "number",
    "items",
    "item",
    "qty",
    "quantity",
    "count",
    "sold",
    "transactions",
    "transaction",
    "pieces",
    "units",
    "lines",
    "savings",
    "saved",
];

/// Return whether joined line text reads as a grand-total row.
pub fn is_grand_total_line(line_text: &str) -> bool {
    if line_text.is_empty() {
        return false;
    }
    if !TOTAL_KEYWORD_RE.is_match(line_text) {
        return false;
    }
    if SUBTOTAL_KEYWORD_RE.is_match(line_text) {
        return false;
    }
    if TAX_KEYWORD_RE.is_match(line_text) {
        return false;
    }
    if NON_PAYMENT_SUMMARY_RE.is_match(line_text) {
        return false;
    }
    let lower = line_text.to_ascii_lowercase();
    !TOKEN_SPLIT
        .split(&lower)
        .filter(|t| !t.is_empty())
        .any(|token| GRAND_TOTAL_DISQUALIFIER_TOKENS.contains(&token))
}

/// Parse receipt currency/amount text without changing the OCR text.
pub fn parse_receipt_amount(text: &str) -> Option<f64> {
    let raw = text.trim();
    if raw.is_empty() {
        return None;
    }

    let mut is_negative = raw.starts_with('(') && raw.ends_with(')');
    let mut cleaned = STRIP_CURRENCY_WS.replace_all(raw, "").into_owned();

    if cleaned.starts_with('(') && cleaned.ends_with(')') {
        cleaned = cleaned[1..cleaned.len() - 1].to_string();
    }

    if cleaned.ends_with('-') {
        is_negative = true;
        cleaned.pop();
    }

    if cleaned.starts_with('-') {
        is_negative = true;
        cleaned.remove(0);
    }

    cleaned = NON_NUMERIC.replace_all(&cleaned, "").into_owned();
    if cleaned.is_empty() || !HAS_DIGIT.is_match(&cleaned) {
        return None;
    }

    let normalized = normalize_decimal_separators(&cleaned)?;
    let value: f64 = normalized.parse().ok()?;
    Some(if is_negative { -value } else { value })
}

/// Return whether text has receipt amount punctuation/symbol context.
pub fn looks_like_receipt_amount(text: &str) -> bool {
    let raw = text.trim();
    if raw.is_empty() || !HAS_DIGIT.is_match(raw) {
        return false;
    }
    CURRENCY_RE.is_match(raw)
        || full_match(&GROUPED_AMOUNT_RE, raw)
        || full_match(&DECIMAL_AMOUNT_RE, raw)
}

fn full_match(re: &Regex, text: &str) -> bool {
    re.find(text)
        .is_some_and(|m| m.start() == 0 && m.end() == text.len())
}

fn normalize_decimal_separators(cleaned: &str) -> Option<String> {
    if !cleaned.contains(',') {
        return Some(cleaned.to_string());
    }

    let last_comma = cleaned.rfind(',')?;
    let last_dot = cleaned.rfind('.');

    if let Some(last_dot) = last_dot {
        if last_comma > last_dot {
            if EURO_GROUPED.is_match(cleaned) {
                return Some(cleaned.replace('.', "").replace(',', "."));
            }
            return Some(cleaned.replace('.', "").replace(',', "."));
        }
        return Some(cleaned.replace(',', ""));
    }

    if THOUSANDS_ONLY.is_match(cleaned) {
        return Some(cleaned.replace(',', ""));
    }
    if DECIMAL_COMMA.is_match(cleaned) {
        return Some(cleaned.replace(',', "."));
    }
    Some(cleaned.replace(',', ""))
}

/// Exposed for tests that assert the Python-mirrored tender regex still compiles.
pub fn has_tender_keyword(text: &str) -> bool {
    TENDER_KEYWORD_RE.is_match(text)
}
